### PR TITLE
DPE-4200 Scale-up from zero

### LINKED ISSUE #409: Mysql-k8s breaks after scale down 0
Issue description:

Using the juju and charms versions below:
```
juju status 
Model  Controller        Cloud/Region      Version  SLA          Timestamp
test   my-k8s-localhost  my-k8s/localhost  3.2.4    unsupported  18:07:02-03:00

App             Version  Status   Scale  Charm           Channel      Rev  Address        Exposed  Message
mysql-k8s                waiting      1  mysql-k8s       8.0/stable   127  10.152.183.28  no       installing agent
mysql-test-app  0.0.2    waiting      1  mysql-test-app  latest/edge   36  10.152.183.94  no       installing agent

Unit               Workload  Agent      Address       Ports  Message
mysql-k8s/0*       unknown   executing  10.1.192.108         
mysql-test-app/0*  waiting   idle       10.1.192.107         
```

After scaling down the mysql-k8s   to 0 and scaling up to 1 the mysql-k8s is broken on juju side, the k8s side shows the pod running:
```
NAME                             READY   STATUS    RESTARTS   AGE
modeloperator-847d5d9b4c-qxdxr   1/1     Running   0          9m26s
mysql-k8s-0                      2/2     Running   0          5m44s
mysql-test-app-0                 1/1     Running   0          8m37s
```
I was able to simulate this issue after having some app with the database relation like the app mysql-test-app-0. 

DEBUG LOG:

```
juju deploy mysql-k8s --trust --channel 8.0/stable


unit-mysql-k8s-0: 18:07:15 ERROR unit.mysql-k8s/0.juju-log database:4: Timed out waiting for k8s service to be ready
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/src/relations/mysql_provider.py", line 124, in _on_database_requested
    self.charm.k8s_helpers.wait_service_ready((primary_endpoint, 3306))
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/tenacity/__init__.py", line 289, in wrapped_f
    return self(f, *args, **kw)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/tenacity/__init__.py", line 379, in __call__
    do = self.iter(retry_state=retry_state)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/tenacity/__init__.py", line 325, in iter
    raise retry_exc.reraise()
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/tenacity/__init__.py", line 158, in reraise
    raise self.last_attempt.result()
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 451, in result
    return self.__get_result()
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/tenacity/__init__.py", line 382, in __call__
    result = fn(*args, **kwargs)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/src/k8s_helpers.py", line 201, in wait_service_ready
    raise TimeoutError
TimeoutError
unit-mysql-k8s-0: 18:07:15 ERROR unit.mysql-k8s/0.juju-log database:4: Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/./src/charm.py", line 770, in <module>
    main(MySQLOperatorCharm)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/ops/main.py", line 456, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/ops/main.py", line 144, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/ops/framework.py", line 351, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/ops/framework.py", line 853, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/ops/framework.py", line 943, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/lib/charms/data_platform_libs/v0/data_interfaces.py", line 1928, in _on_relation_changed_event
    getattr(self.on, "database_requested").emit(
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/ops/framework.py", line 351, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/ops/framework.py", line 853, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/ops/framework.py", line 943, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/src/relations/mysql_provider.py", line 124, in _on_database_requested
    self.charm.k8s_helpers.wait_service_ready((primary_endpoint, 3306))
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/tenacity/__init__.py", line 289, in wrapped_f
    return self(f, *args, **kw)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/tenacity/__init__.py", line 379, in __call__
    do = self.iter(retry_state=retry_state)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/tenacity/__init__.py", line 325, in iter
    raise retry_exc.reraise()
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/tenacity/__init__.py", line 158, in reraise
    raise self.last_attempt.result()
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 451, in result
    return self.__get_result()
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/venv/tenacity/__init__.py", line 382, in __call__
    result = fn(*args, **kwargs)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/src/k8s_helpers.py", line 201, in wait_service_ready
    raise TimeoutError
TimeoutError
unit-mysql-k8s-0: 18:07:15 ERROR juju.worker.uniter.operation hook "database-relation-changed" (via hook dispatching script: dispatch) failed: exit status 1
unit-mysql-k8s-0: 18:07:15 INFO juju.worker.uniter awaiting error resolution for "relation-changed" hook
unit-mysql-k8s-0: 18:07:25 INFO juju.worker.uniter awaiting error resolution for "relation-changed" hook
unit-mysql-k8s-0: 18:07:26 ERROR unit.mysql-k8s/0.juju-log database:4: Failed to get cluster status for cluster-7f7c9a3504e4f58a371acc55cca28b49
unit-mysql-k8s-0: 18:07:26 ERROR unit.mysql-k8s/0.juju-log database:4: Failed to get cluster endpoints
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/src/mysql_k8s_helpers.py", line 836, in update_endpoints
    rw_endpoints, ro_endpoints, offline = self.get_cluster_endpoints(get_ips=False)
  File "/var/lib/juju/agents/unit-mysql-k8s-0/charm/lib/charms/mysql/v0/mysql.py", line 1469, in get_cluster_endpoints
    raise MySQLGetClusterEndpointsError("Failed to get endpoints from cluster status")
charms.mysql.v0.mysql.MySQLGetClusterEndpointsError: Failed to get endpoints from cluster status
unit-mysql-k8s-0: 18:07:26 WARNING unit.mysql-k8s/0.juju-log database:4: Kubernetes service already exists
```

So we need a way to fix the juju status.

### PR DESCRIPTION
**Pull request by paulomach**
## Issue

O scaling to zero, the last unit ensure the cluster is dissolved.
Scale back to unit > 0 don't detect and treat the dissolved cluster.

## Solution

* detect and treat special case of dissolved cluster

Fixes #409

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -826,6 +857,11 @@ def _on_database_storage_detaching(self, _) -> None:
         # Inform other hooks of current status
         self.unit_peer_data["unit-status"] = "removing"
 
+        if self.unit.is_leader():
+            # Update 'units-added-to-cluster' counter in the peer relation databag
+            units = int(self.app_peer_data.get("units-added-to-cluster", 1))
+            self.app_peer_data["units-added-to-cluster"] = str(units - 1)
```
Comment (shayancanonical) [inline src/charm.py:863]: should we also have this update to `units-added-to-cluster` on the `peer-relation-departed` event? wouldnt we like to decrement this value if non-leader units are scaled down?

or is this happening on `update-status` (in particular `_set_app_status`)? if this is the case, would there be inconsistency issues -> there are 3 units, if another unit departed and then the leader unit departed, `units-added-to-cluster` will be 2 instead of 1 (if `storage-detaching` runs on leader before `update-status`)

File: src/charm.py [hunk 2]
```diff
@@ -252,6 +252,20 @@ def unit_address(self) -> str:
...
+
+        e.g. scaling from zero units
+        """
+        _default_unit_data_keys = {
+            "egress-subnets",
+            "ingress-address",
+            "private-address",
+            "unit-status",
+        }
+        return self.unit_peer_data.keys() == _default_unit_data_keys
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/charm.py:267]: can we utilize `units-added-to-cluster` here instead of relying on these keys?

[the remark this answers is not in the document]
Comment (taurus-forever) [reply src/charm.py:267]: IMHO, we can rely only on data on the storage (it can be foreign storage by user misatke, which we should not damage/erase).

File: src/charm.py [hunk 3]
```diff
@@ -632,6 +644,20 @@ def _on_mysql_pebble_ready(self, event) -> None:
             # Data directory is already initialised, skip configuration
             logger.debug("Data directory is already initialised, skipping configuration")
             self._reconcile_pebble_layer(container)
+            if self.is_new_unit:
+                # when unit is new and has data, it means the app is scaling out
+                # from zero units
+                logger.debug("Scaling out from zero units")
+                if self.unit.is_leader():
+                    # create the cluster due it being dissolved on scale-down
+                    self.create_cluster()
```
[the remark this answers is not in the document]
Comment (taurus-forever) [reply src/charm.py:653]: IMHO, charm must be blocked if foreign disk attached by mistake (to avoid data damage).

Today it is not possible for K8s, but Pedro is working with Juju on it:
```
> juju scale-application mysql 0
mysql scaled to 0 units

> juju add-storage mysql/0 database=foreigndisk
ERROR Juju command "add-storage" not supported on container models    # will be supported on K8s like on VM

> juju scale-application mysql 3
...
```



----

Another reason to write to disk:
```
> juju scale-application mysql 0

# juju controller crashed... restored ...

> juju scale-application mysql 1  # did we loose units-added-to-cluster ?
```

**Comment by taurus-forever** [comment]
BTW, we are adding the test to check scale-down to zero and restore with foreign disk [here](https://github.com/canonical/postgresql-k8s-operator/pull/509).

Maybe it worth to copy it into MySQL VM/K8s as well?

**Comment by paulomach** [comment]
> we might need to change some things on router to support scaling mysql-k8s to 0 and back up

Testing it, it requires a remove and re-relate. On scaling in to zero, routers are cleaned up from the metadata already.
Messaging in the router is very clear, imo I think that's adequate behavior.

**Comment by taurus-forever** [comment]
> re:  `it requires a remove and re-relate`

Fields/customers do not like this... can we avoid re-relations without the real over-complication in code?

**Comment by sombrafam** [comment]
@paulomach hi Paulo, I see that the pool request was merged. How can I know the release version this will be published if it's not already?

**Comment by paulomach** [comment]
Hey @sombrafam , this still on the discussion. There are some edge cases we need to fix.
